File: ws/src/demo_basic_cpp/src/talker.cpp
// Demo 1 — minimal rclcpp publisher ("talker_cpp"), the C++ side for a firmware audience.
// Publishes on the SAME /chatter topic as the Python talker: one bus, two languages, no master.
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class TalkerCpp : public rclcpp::Node
{
public:
  TalkerCpp()
  : Node("talker_cpp"), count_(0)
  {
    pub_ = create_publisher<std_msgs::msg::String>("chatter", 10);
    timer_ = create_wall_timer(1s, [this]() {this->tick();});   // 1 Hz
    RCLCPP_INFO(get_logger(), "talker_cpp up: publishing on /chatter");
  }

private:
  void tick()
  {
    auto msg = std_msgs::msg::String();
    msg.data = "Hello from C++ (rclcpp): " + std::to_string(count_++);
    pub_->publish(msg);
    RCLCPP_INFO(get_logger(), "pub: \"%s\"", msg.data.c_str());
  }

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  size_t count_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TalkerCpp>());
  rclcpp::shutdown();
  return 0;
}

File: ws/src/demo_basic/demo_basic/talker.py
#!/usr/bin/env python3
"""Demo 1 — minimal rclpy publisher ("talker").

Firmware analogy: a node appearing on the ROS 2 graph is like a device enumerating on a
shared bus — no master assigns it an address, discovery is automatic. This talker just
publishes a counter string on /chatter at 2 Hz.
"""
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class Talker(Node):
    def __init__(self):
        super().__init__('talker')                          # name shown in `ros2 node list`
        self.pub = self.create_publisher(String, 'chatter', 10)  # topic, queue depth 10
        self.timer = self.create_timer(0.5, self.tick)      # 2 Hz
        self.count = 0
        self.get_logger().info('talker up: publishing on /chatter')

    def tick(self):
        msg = String()
        msg.data = f'Hello ROS 2 (rclpy): {self.count}'
        self.pub.publish(msg)
        self.get_logger().info(f'pub: "{msg.data}"')
        self.count += 1


def main(args=None):
    rclpy.init(args=args)
    node = Talker()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()
